pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+        session_idx: SessionIndex,
+    }
+
+    #[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
+    /// The activity status of the validator
+    pub enum ValidatorStatus {
+        /// Committed to be online and producing valid blocks
+        Active,
+        /// Temporarily inactive
+        Idle,
+        /// Bonded until the inner round
+        Leaving(SessionIndex),
+    }
+
+    impl Default for ValidatorStatus {
+        fn default() -> ValidatorStatus {
+            ValidatorStatus::Active
+        }
+    }
+
+    #[derive(Encode, Decode, Clone, RuntimeDebug)]
+    /// Global validator state with commission fee, bonded stake, and nominations
+    pub struct Validator<AccountId, Balance> {
+        pub id: AccountId,
+        pub bond: Balance,
+        pub nomi_bond_total: Balance,
+        pub nominators: OrderedSet<Bond<AccountId, Balance>>,
+        pub total: Balance,
+        pub state: ValidatorStatus,
+        pub unlocking: Vec<UnlockChunk<Balance>>,
+    }
+
+    impl<
+            A: Ord + Clone,
+            B: AtLeast32BitUnsigned + Ord + Copy + sp_std::ops::AddAssign + sp_std::ops::SubAssign,
+        > Validator<A, B>
+    {
+        pub fn new(id: A, bond: B) -> Self {
+            let total = bond;
+            Validator {
+                id,
+                bond,
+                nomi_bond_total: Zero::zero(),
+                nominators: OrderedSet::new(),
+                total,
+                state: ValidatorStatus::default(), // default active
+                unlocking: Vec::new(),
+            }
+        }
+        pub fn is_active(&self) -> bool {
+            self.state == ValidatorStatus::Active
+        }
+        pub fn is_leaving(&self) -> bool {
+            matches!(self.state, ValidatorStatus::Leaving(_))
+        }
+        pub fn bond_more(&mut self, more: B) {
+            self.bond += more;
+            self.total += more;
+        }
+        // Returns None if underflow or less == self.bond (in which case validator should leave)
+        pub fn bond_less(&mut self, less: B) -> Option<B> {
+            if self.bond > less {
+                self.bond -= less;
+                // self.total -= less;
+                Some(self.bond)
+            } else {
+                None
+            }
+        }
+        pub fn inc_nominator(&mut self, nominator: A, more: B) {
+            for x in &mut self.nominators.0 {
+                if x.owner == nominator {
+                    x.amount += more;
+                    self.nomi_bond_total += more;
+                    self.total += more;
+                    return;
+                }
+            }
+        }
+        pub fn dec_nominator(&mut self, nominator: A, less: B) {
+            for x in &mut self.nominators.0 {
+                if x.owner == nominator {
+                    x.amount -= less;
+                    self.nomi_bond_total -= less;
+                    self.total -= less;
```

Comment:
Same comment for checked math or saturating operators

Reply:
:+1:  Moved to saturating_sub()